### PR TITLE
Update lip-8

### PR DESCRIPTION
**Pull request by danprinz**
Disclaimer regarding the discovery phase
Extended scope for fund pull pre-approval
Revoke an authorization in more details

### CODE REVIEW

**Comment by JoelMarcey** [comment]
@danprinz Looks like a rebase may be needed.

@bmwill Is there something we need to do with `bors` here? I don't usually see that warning in PRs to this repo, I don't think.

**Comment by bmwill** [comment]
@JoelMarcey Did we ever  resolve if we wanted bors on this repo or not? If not I probably need to just disable bors since IIRC we aren't currently relying on bors for this repo.

**Comment by JoelMarcey** [comment]
@bmwill I think I am ok removing `bors` on this repo for now and revisit later. We have the passive checks to make sure the website builds and we have only `CODEOWNERS` being able to merge, which helps too.

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:30]: I think we should leave this portion out.  It's outside the scope of this LIP and I think it's enough to mention that it isn't in the scope of this LIP without speculating how it may be done

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:146]:
```suggestion
In this object the initiator VASP declares its intent for the pre-approval, this can be one of two options:
```

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:149]:
```suggestion
Also, the scope limits the FundPullPreApprovalObject to certain parameters of time and amount. This object can be later mutated by the initiator VASP if needed, but any change requires the target VASP to approve the scope change.
```

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:154]:
```suggestion
| type | str enum | Y | This can be either save_sub_account or consent |
```

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:157]:
```suggestion
| max_transaction_amount | [CurrencyObject](#currencyobject) | N | Max transaction amount that is approved for funds pre-approval.  This is the maximum amount that may occur in a single transaction while utilizing this funds pre-approval. |
```

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:161]: Let's add the max_cumulative_amount to this also

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:58]: This doesn't have "type" in it

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:154]: I'm not sure I understand why we would want "type".  save_sub_account seems to not be overly useful, but perhaps I'm missing something.  Isn't it implied that if I wanted to bill someone again, I could try their same subaddress and they would have the option to approve it?  Even if I didn't get this approval ahead of time, I could still trigger the transaction, couldn't I?

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:178]: Should we add "day"?  And what about something for an all-time limit - or should a high number of years suffice for that?

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:179]:
```suggestion
| value         | int        | N            | Number of "unit"(s)   |
```

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:291]:
```suggestion
The second (authorization) could be canceled only by the biller (merchant) as it holds a guarantee of funds availability if requested. This has no reason to be canceled by the buyer. a request for such cancellation from the buyer should be rejected. 
```

**Review by kphfb — CHANGES_REQUESTED** [review]

File: lips/lip-8.md
Comment (danprinz) [reply lips/lip-8.md:178]: for "days" - you're right. I will add it. 
for "all-time" - I would go with the latter to keep the structure the same for all options (if we set a unit value of "all-time" then we don't need a "value" field which makes parsing uglier). not crucial, just a matter of taste.

File: lips/lip-8.md
Comment (danprinz) [reply lips/lip-8.md:30]: Agreed. I will remove it.

File: lips/lip-8.md
Comment (danprinz) [reply lips/lip-8.md]: you are right that "consent" covers the basic option of "save_sub_account". 
but we thought it would be nice to have the user approve even the storage of their details (sub-account). 
consider the scenario of users who wants to add their Libra wallet as a payment method (to make subsequent payments easier), but still want to authorize every payment. something like a push notification to their mobile app. 
the merchant VASP should store their sub-address but without a "real" consent of withdrawing funds. 
one may argue that the merchant could save these details without prior approval of the user, but I think that users will want to authorize even this. 
the wallet VASPs on the other hand will pop a notification to the user only if the merchant could present such an approval.    

does it make sense?

File: lips/lip-8.md
Comment (dahliamalkhi) [inline lips/lip-8.md:26]: I am not sure we want to include any statements here on discovery methods, since this is an evolving area and there is currently no method(s) that Libra has standardized. I would rephrase and say that the industry has mechanisms such as [...] and others.

File: lips/lip-8.md
Comment (dahliamalkhi) [inline lips/lip-8.md:32]: I am missing a high level description of the proposed approach, before we jump into the Specification.

**Review by dahliamalkhi — CHANGES_REQUESTED** [review]

File: lips/lip-8.md
Comment (dahliamalkhi) [inline lips/lip-8.md:21]: Now that Libra 1.0.0 is released, this LIP should not be Informational but Standard.

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:30]: Yeah, I think we should just leave this out since it's speculation

File: lips/lip-8.md
Comment (kphfb) [reply lips/lip-8.md]: Hmmm, I'm still not convinced that this is something that merchants/users will want.  I could be wrong though.  But it feels like there is no real reason for anyone to follow this for save_sub_account.  I can just put a checkbox on my page to "save" in the same way and it functionally operates exactly the same as this without needing any off-chain or user action in their wallet app.  It just feels like we are adding a feature that isn't actually adding anything practically, but does add some complexity in implementation

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:229]: This would still be billee.  Or maybe we should user "payer" and "payee" instead.  In which case, this would be the payer VASP.  The side that has the user on it.  Actually, let's just call it the sender VASP (buyer) maybe

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:256]:
```suggestion
When an authorization happens, the sender VASP (buyer) who agrees to the authorization request must lock the funds for the specified period. That is - the buyer VASP guarantees that the funds will be available if later captured.
```

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:298]:
```suggestion
The first one (pre-approval) may be canceled by either side (buyer or seller). A reasonable scenario is when a consumer wishes to cancel a subscription (buyer cancel) or asks a merchant app to remove the user wallet from the list of payment methods (seller cancel). 
```

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:300]:
```suggestion
The second (authorization) could be canceled only by the biller (merchant) as it holds a guarantee of funds availability if requested. This has no reason to be canceled by the buyer. A request for such cancellation from the buyer should be rejected. 
```

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:122]:
```suggestion
The structure in this object can be a full pre-approval or just the fields of an existing pre-approval object that need to be changed. Some fields are immutable after they are defined once (see below). Others can be updated multiple times. Updating immutable fields with different values results in a command error, but it is acceptable to re-send the same value. It should be noted that initial creation of the FundPullPreApprovalObject can be created or updated by the VASP on either side (buyer or seller).
```

**Review by kphfb — CHANGES_REQUESTED** [review]

**Comment by kphfb** [comment]
Think with the one change I suggest, it hopefully addresses https://github.com/libra/lip/issues/71 as well.  If so, we can close that one once this is updated

File: lips/lip-8.md
Comment (kphfb) [inline lips/lip-8.md:26]:
```suggestion
For now, we will assume that such a discovery process took place and the merchant has the buyer subaddress at hand.
```

**Comment by kphfb** [comment]
I think I'm mostly good with this PR other than I'm still not convinced about "type" and want to chat about that first

File: lips/lip-8.md
Comment (kphfb) [reply lips/lip-8.md:29]: @dahliamalkhi what type of thing would you like to see here as compared to what is there in the abstract section?

File: lips/lip-8.md
Comment (LuZhang-Lou) [inline lips/lip-8.md:289]: after reading the whole auth/capture section, I still do not understand how the valid status transition. Can we have a detailed diagraph like we are making for lip-1?

File: lips/lip-8.md
Comment (xli) [inline lips/lip-8.md:147]: what is "sub_account" means?
use name "type" is confusing here, as type indicates the object type, which is protocol level,
but it looks like a business agreement about how the next payment object can be processed.

File: lips/lip-8.md
Comment (xli) [inline lips/lip-8.md:185]: what is difference between above 2 fields with "expiration_timestamp"? they are also confusing because different application may have different interpretation about what means of start of the day/week/month/year, hence not accurate.

Comment (danprinz) [reply lips/lip-8.md:185]: These fields are for limiting (scoping) the fund pulls from the consumer VASP. It can't be expressed using an expiration time, as it might be "not more than $100 a week". 
Regarding the interpretation of week start, duration, etc. - we think of it as a sliding time window, but one may think it's a calendar timeframe.
I will add a more detailed description of it.

File: lips/lip-8.md
Comment (danprinz) [reply lips/lip-8.md:125]: It won't be very clear once one will try to use it in a subsequent payment (a reference to the user consent). 
IMHO this name is more descriptive.

File: lips/lip-8.md
Comment (danprinz) [reply lips/lip-8.md:147]: "sub_account" is the consumer identifier within (behind) the VASP. 
It might be a good idea to replace it with another term that indicates the wallet user.

File: lips/lip-8.md
Comment (davidiw) [reply lips/lip-8.md:26]: I think the normal approach would be to describe a general strawman flow, then explicitly state what is assumed (e.g., that there is a user -> merchant identity exchange process). As I noted in another PR on DIP-8, there's so much boiler plate before getting into what this DIP does, that it ... isn't particularly accessible. You calling this out, means you see it too :).

File: lips/lip-8.md
Comment (davidiw) [inline lips/lip-8.md:35]: Here's another assumption, an out-of-band process, I think having a strawman definition of this protocol broken down into bullets would be more accessible. We could even allude to one such method for doing the exchanges.

File: lips/lip-8.md
Comment (davidiw) [inline lips/lip-8.md:55]: what's the value of having another data structure? could we amend the PR description or is there a tracking doc for this?

File: lips/lip-8.md
Comment (davidiw) [inline lips/lip-8.md:119]: This seems kinda weird to me, it feels like it would make more sense for this to be something like a client submits a pre-approval to the merchant's VASP. There shouldn't be any mutable fields either it is agreeable or not. If it is not, return back some error that the service can then address. If the services should support follow up pre-approvals, those should be separate interactions not building on the same "object". I should have asked this on Kevin's PR, but I'll ask it on yours instead.

File: lips/lip-8.md
Comment (davidiw) [reply lips/lip-8.md:125]: I think the key thing here is .... how do we distinctly distinguish different objects... a reference_id is supposed to be common to all objects -- if there is a shared object to be created or maintained. Also it might behove us to pivot to DiemId rather than further embracing the jankiness of subaddresses, they are really dizzying.

File: lips/lip-8.md
Comment (davidiw) [reply lips/lip-8.md:147]: ahh, interesting insight, we should determine a uniform fashion to call these out. I am concerned that subaccount doesn't really refer to a client's account and I've heard a lot of confusion across subaddress and child vasps :/

File: lips/lip-8.md
Comment (davidiw) [reply lips/lip-8.md:185]: unit - reset_period?
value - amount_per_period

also I would probably move max_amount and currency to the same level since currency refers to both...

File: lips/lip-8.md
Comment (davidiw) [inline lips/lip-8.md:297]: shouldn't there be a follow up command to cancel?

I think having a clean story of how these pieces fit together would be immensely valuable for this DIP. As a super noob to the space, I'm struggling a bit to see how all these fit together.

**Review by davidiw — COMMENTED** [review]
Really excited to see the enthusiasm in the space! And it makes it painfully difficult how hard it can be to work effectively in github for doc related reviews :P

I think the biggest contribution here is to enable the notion of periodic resettable limits within a single approval --- that makes some sense. I'm not aware of whether or not this is a feature that existing payment providers offer, so would we potentially be increasing the barrier to entry? With that being said, I do like the concept and am wondering if it should be a different command or a follow up.

**Comment by davidiw** [comment]
feel free to re-open but this is pretty stale now

**Pull request closed without merging**
